### ISSUE TITLE
MappingR2dbcConverter always calls Row::get(String) instead Row::get(String,Class)

### ISSUE THREAD

**Issue by Bittuw** [spring-projects/spring-data-r2dbc#502]
[In R2DBC](https://github.com/r2dbc/r2dbc-h2/issues/177) as example of obscured types.
Is it normally that `MappingR2dbcConverter` doesn't call `Row::get(String, Class)`? I thought that `MappingR2dbcConverter` provides all target field types transformations to r2dbc-spi implementation. But instead it calls `Row::get(String)` and SPI implementation doesn't know about the target field type.

**Comment by mp911de** [comment]
Yes, that's by intent. There's a bit of background to understand why we don't use the target type.

Spring Data is a library that ships with extensive mapping and conversion abstractions. Spring Data can map various types to the domain model applying lenient conversion. Using `Row.get(…)` with a `Class` parameter requires the target type to be a R2DBC primitive and it requires that a driver can convert the actual type into the desired target type. 

Drivers typically don't convert timestamps or numeric values to e.g. `String`. Going forward, Spring Data allows using converters to convert the actual database value into something else (e.g. for JSON `String` to `YourPersonObject`). While Spring Data is aware of how to convert a JSON `String` into a complex object, the R2DBC is not.

You seem to have an issue with the current approach and we'd like to understand your use-case, what you want to achieve and which obstacles you encounter.

**Comment by Bittuw** [comment]
@mp911de, Thanks for explaning. My problem:
As i mentioned in linked issue, I use `Duration` field type in my DTOs and H2 as default DB. If I don't override any converters in Spring Data R2dbc, `Duration` class will be use the next mapping chain:
1. **Reading**: H2 will be use `IntervalCodec` because its doesn't know about field type and `Interval` is assignable to Object class (Which is default). So there will be convertion error, because there is no any converter from `Interval` to `Duration`;
2. **Writing**:  The first, spring will convert `Duration` into `String` (I don't know why Spring Data R2dbc provide this converter by default), then there will be invalid SQL bindings error, because we cannot assign `String` to `Interval` in SQL.

To avoid this behavior I use special writing `Converter<Duration,Duration>` to override `Converter<Duration,String>` and reading `Converter<Interval, Duration>`.

I think that its very strange behavior, because H2 r2dbc driver can convert `Duration`  itself and it will be useful to "ask" R2DBC driver implementation, that it can convert this type itself before using any converters from `ConverterService`.

**Comment by mp911de** [comment]
`Converter<Duration,String>` comes from Spring Data's early days JSR-310 support where `Duration` were converted to `String` because JPA, MongoDB, and other stores didn't support `Duration`.

Generally speaking, `Intervall` is an H2 native type. Writing-wise, it could make sense to register a converter `Duration` to `Interval.ofDaysHoursMinutesNanos(…)` and `Interval` to `Duration`.

Out of interest, how would that work with JDBC?

**Comment by Bittuw** [comment]
But what about the layered transformations I suggested? It is more flexible way to support already implemented convertions in H2/MySQL/Postgres and not implemented JPA/MongoDB. Actually `Row::get(String)` can be replaced by:
``` java
try { return row.get("table_name", field_type); }
catch (CannotConvert e) { /*logs*/ }
return convertionService.convert(row.get("table_name"), field_type);
```
What to do with `Converter<Duration,String>` i don't know, because I don't know who provide this converter? R2DBC SPI implementation or Spring Data R2DBC itself?

**Comment by mp911de** [comment]
Exception-driven conditionals cause a lot of performance drag for an edge-case that could be solved in a better style.

Regarding converters, we can ship these with Spring Data R2DBC.

**Comment by Bittuw** [comment]
@mp911de , not necessarily as I wrote. I think it's enough to be able to ask R2DBC SPI implementation about possability of such convertion. But how it must be done I don't know.
